Fixes #13228 - Update Core Deploy Documentation. (#13322)
eclipse/jetty.project: documentation/jetty/modules/operations-guide/pages/deploy/index.adoc

= Web Application Deployment

The Jetty server supports the deployment of these types of web applications:

* <<web-application-format-static,Jetty Static web applications>> -- web applications made of static files only.
* <<web-application-format-core,Jetty Core web applications>> -- web applications written against the Jetty `Handler` APIs, without any Servlet dependencies.
* <<web-application-format-jakarta,Java EE 8 web applications>> -- web applications written using the Servlet 4.0 (and associated specifications) in the `+javax.*+` packages.
* <<web-application-format-jakarta,Jakarta EE 9 to Jakarta {ee-current-caps} web applications>> -- web applications written using the corresponding Servlet (and associated specifications) in the `+jakarta.*+` packages.

> In the following, the term "web application" is used to refer to either a Jetty Static web application, or Jetty Core web application, or Java EE 8 web application or Jakarta EE 9 to Jakarta {ee-current-caps} web application, unless specified otherwise.

Reviewer: When coming up with the coreapp vs webapp vs staticapp naming convention, we explicitly said that we would not use "web application" in the generic sense as it overloads that name in a confusing way and if somebody says "deploy the web application" we have no idea what they mean.

Let's pick a term for the generic.   It could just be "application", but I still don't mind "context".  Whilst "context" is somewhat overloaded, we do know that every "application" has a `ContextHandler` that is configured with a `contextPath` and that requests entering the "application" are given a `Context` instance that describes the "application" and that for web applications that context is a `ServletContext`.  It is all context!

Reviewer 2: I agree with @gregw . We specifically used the terms `coreapp`, `webapp` and `staticapp`. We cannot not now refer to all of them as web applications without nullifying the clear delineation of the types of application. So IMHO they should be referred to as a group as `applications`, which is a widely understood term, and fits nicely with the specialized cases of `core application`, `web application` and `static application`.

> The Jetty server supports the deployment of these types of web applications:
> 
> * <<web-application-format-static,Jetty Static web applications>> -- web applications made of static files only.
> * <<web-application-format-core,Jetty Core web applications>> -- web applications written against the Jetty `Handler` APIs, without any Servlet dependencies.
> * <<web-application-format-jakarta,Java EE 8 web applications>> -- web applications written using the Servlet 4.0 (and associated specifications) in the `+javax.*+` packages.
> * <<web-application-format-jakarta,Jakarta EE 9 to Jakarta {ee-current-caps} web applications>> -- web applications written using the corresponding Servlet (and associated specifications) in the `+jakarta.*+` packages.
> 
> In the following, the term "web application" is used to refer to either a Jetty Static web application, or Jetty Core web application, or Java EE 8 web application or Jakarta EE 9 to Jakarta {ee-current-caps} web application, unless specified otherwise.

Reviewer: We can also deploy directly configured `ContextHandlers` with a context XML file:

```suggestion
The Jetty server supports the deployment of the following type of web applications:

* <<web-application-format-jakarta,Jakarta EE 9 to Jakarta {ee-current-caps} web applications>> -- web applications written using the corresponding Servlet (and associated specifications) in the `+jakarta.*+` packages.
* <<web-application-format-jakarta,Java EE 8 web applications>> -- web applications written using the Servlet 4.0 (and associated specifications) in the `+javax.*+` packages.
* <<web-application-format-core,Jetty Core web applications>> -- web applications written against the Jetty `Handler` APIs in a Jetty specific deployment structure, without any Servlet dependencies.
* <<web-application-format-static,Jetty Static web applications>> -- web applications made of static files only.
* <<web-application-context-handler,Jetty Context Handler>> -- web application made by an XML configured `ContextHandler` instance.

In the following, the term "web application" is used to refer to either a Jetty Static web application, or Jetty Core web application, or directly configured `ContextHandler`, or Java EE 8 web application or Jakarta EE 9 to Jakarta {ee-current-caps} web application, unless specified otherwise.
```

Reviewer 2: @joakime if we use "web application" generically, then everywhere where we have documentation that is specific to "ee web applications" we need to edit it and fix.

Author: @janbartel I have done that.

Of course, I might have missed a few, feel free to point them out.

Also, in the following, references to "Jakarta EE" refer to both Java EE and Jakarta EE, unless specified otherwise.

The Jetty server deploys each web application to a specific _environment_.

The available environments are:

> * The `static` environment, for Jetty Static web applications.

Reviewer 2: ```suggestion
* The `static` environment, for Jetty Static applications.
```

Reviewer: We have agreed on "web application", so I'll resolve for clarity.

> * The `core` environment, for Jetty Core web applications.

Reviewer 2: ```suggestion
* The `core` environment, for Jetty Core applications.
```

> * The `core` environment, for Jetty Core web applications.

Reviewer: ```suggestion
* The `core` environment, for Jetty Core web applications and XML configured `ContextHandler`s.
```

* The `{ee-all}` environments, for Jakarta EE web applications.

> You can simultaneously deploy an old Java EE 8 web application, say `old-ee8.war`, alongside a new Jakarta {ee-current-caps} web application, say `new-{ee-current}.war`, alongside a web application that only uses the Jetty Core `Handler` APIs, say `backoffice-core.xml`, all in the same Jetty server.

Reviewer 2: ```suggestion
You can simultaneously deploy an old Java EE 8 web application, say `old-ee8.war`, alongside a new Jakarta {ee-current-caps} web application, say `new-{ee-current}.war`, alongside an application that only uses the Jetty Core `Handler` APIs, say `backoffice-core.xml`, all in the same Jetty server.
```

* The Java EE 8 Environment -- Supports Servlet 4.0 (and associated specifications) in the `+javax.*+` packages.
* Any version from the Jakarta EE 9 Environment up to the Jakarta {ee-current-caps} Environment -- Supports corresponding Servlet (and associated specifications) in the `+jakarta.*+` packages.

You can simultaneously deploy an old Java EE 8 web application, say `old-ee8.war`, alongside a new Jakarta {ee-current-caps} web application, say `new-{ee-current}.war`, alongside a web application that only uses the core Jetty `Handler` APIs, all in the same Jetty server.

For each specific environment there is a specific deploy module that must be enabled:

* For Jetty Static, the xref:modules/standard.adoc#static-deploy[`static-deploy`] module

> * For Jetty Core, the xref:modules/standard.adoc#core-deploy[`core-deploy`] module

Reviewer: You need to refer to "Core App" or "Core Application" rather than just "Core" everywhere in this document.    You should probably also use "Web Application" more in relation to EExx deployment, so we establish the naming better.

```suggestion
* For Jetty Core App, the xref:modules/standard.adoc#core-deploy[`core-deploy`] module
```

Author: No, I am talking about the environments here, not the web applications.
I'll clarify that.

[[monitored-directories]]
== Monitored Directories

> The `deployment-scanner` module scans the web applications deploy directory specified by the `jetty.deploy.webappsDir` property, by default `$JETTY_BASE/webapps`, for web applications to deploy.

Reviewer: ```suggestion
The `deployment-scanner` module scans the web applications deploy directory specified by the `jetty.deploy.webappsDir` property, by default `$JETTY_BASE/webapps`, for web applications and core applications to deploy.
```

OR

```suggestion
The `deployment-scanner` module scans the web applications deploy directory specified by the `jetty.deploy.webappsDir` property, by default `$JETTY_BASE/webapps`, for contexts to deploy.
```

You might want to establish this language somewhere that a deployable context is a ContextHandler that could be a Webapp, coreapp, staticapp or directly configured core ContextHandler.

Author: I rewrote the initial section, hopefully clearer now, and specified that "web application" refers to any kind -- static, core, javaee and jakartaee.

[[monitored-directories]]
== Monitored Directories

> The `deployment-scanner` module scans the web applications deploy directory specified by the `jetty.deploy.webappsDir` property, by default `$JETTY_BASE/webapps`, for web applications to deploy.

Reviewer 2: ```suggestion
The `deployment-scanner` module scans the applications deploy directory specified by the `jetty.deploy.appsDir` property, by default `$JETTY_BASE/webapps`, for applications to deploy.
```

== Web Application Format

[[web-application-format-static]]

> === Jetty Static Web Application Format

Reviewer 2: ```suggestion
=== Jetty Static Application Format
```

----

[[web-application-format-core]]

> === Jetty Core Web Application Format

Reviewer 2: ```suggestion
=== Jetty Core Application Format
```

> For Jetty Core web applications, the web application format is a directory with the following structure:

Reviewer 2: ```suggestion
For Jetty Core applications, the web application format is a directory with the following structure:
```

This means that files in the directory root, or under any other directory that is not `static/`, are hidden and if requested they will result in a `404 Not Found` response.
If the `static/` directory is missing, then no static files can be served, and therefore all requests are processed dynamically by the ``Handler``s configured in the web application.

Since static files can only be served from the `static/` directory, the `classes/` and `lib/` directories are implicitly hidden, as well as any other directory such as `includes/` or `fragments/`.

The `classes/` directory stores the web application's `+*.class+` files, while the `lib/` directory stores the web application's libraries, typically `+*.jar+` files.
Both these directories are optional.

The `jetty-web.xml` file is an optional in-application <<jetty-web-xml,Jetty Context XML file,>> that allows you to configure the web application deployment.

[[web-application-format-jakarta]]
=== Jakarta EE Format

For Jakarta EE web applications, the deployment unit is the usual `+*.war+` file or a directory with a `+*.war+` structure.

> Like for the <<web-application-format-core,Jetty Core Format>>, the `+*.war+` file or a directory can contain an optional in-application <<jetty-web-xml,Jetty Context XML file>> under the `WEB-INF` directory, that allows you to configure the web application deployment.

Reviewer: ```suggestion
Like for the <<web-application-format-core,Jetty Core Application Format>>, the `+*.war+` file or a directory can contain an optional in-application <<jetty-web-xml,Jetty Context XML file>> under the `WEB-INF` directory, that allows you to configure the web application deployment.
```

│   └── com
│       └── acme
│           └── AcmeHandler.class
├── lib            # Where web application Java libraries reside.
│   └── acme-util.jar

> └── jetty-web.xml  # Web application specific configuration.

Reviewer 2: This file is now badly named. It is nothing to do with `web.xml` but rather another Jetty XML context XML file that is processed after `web.xml`.

Reviewer 2: Also, how are you arranging for this to be processed for a core application, where there is no concept of `Configuration`s  that would read such a file?

Author: @janbartel this is done by `CoreAppContext` directly.

----

The structure is similar to that of a `+*.war+` file.

A notable difference with `+*.war+` files is that static files are only served from the `static/` directory (with the `static/` prefix removed).
This means that files in the directory root, or under any other directory that is not `static/`, are hidden and if requested they will result in a `404 Not Found` response.

> If the `static/` directory is missing, then no static files can be served, and therefore all requests are processed dynamically by the ``Handler``s configured in the web application.

Reviewer 2: ```suggestion
If the `static/` directory is missing, then no static files can be served, and therefore all requests are processed dynamically by the ``Handler``s configured in the application.
```

Since static files can only be served from the `static/` directory, the `classes/` and `lib/` directories are implicitly hidden, as well as any other directory such as `includes/` or `fragments/`, and any file in the directory root.

> The `classes/` directory stores the web application's `+*.class+` files, while the `lib/` directory stores the web application's libraries, typically `+*.jar+` files.

Reviewer 2: ```suggestion
The `classes/` directory stores the application's `+*.class+` files, while the `lib/` directory stores the application's libraries, typically `+*.jar+` files.
```

Both these directories are optional.

> The <<jetty-web-xml,`jetty-web.xml`>> file is an optional in-application Jetty Context XML file, that allows you to configure the web application deployment.

Reviewer 2: ```suggestion
The <<jetty-web-xml,`jetty-web.xml`>> file is an optional in-application Jetty Context XML file, that allows you to configure the application deployment.
```

Properties specified in environment properties file (described in <<environment-configuration,this section>>) and in the web application properties file (described in <<web-application-files,this section>>) are available when processing the Jetty context XML file.

> Remember that Jetty Context XML files in the `$JETTY_BASE/webapps` directory can only refer to classes present in the environment class-path or module-path, and in the server class-path or module-path.

Reviewer 3: ```suggestion
Remember that Jetty Context XML files in the `$JETTY_BASE/webapps` directory can only refer to classes present in the environment class-path (or module-path), and in the server class-path (or module-path).
```

Reviewer 3: This section goes into far more detail than the similar (new) section below talking about com.acme.MyListenerFeature, perhaps we need a link (in both of these places) to new section on how classes are found (the environment classloader, the classpath, the modulepath, etc)

Author: I added a link pointing to the class loader hierarchy section.

Properties specified in environment properties file (described in <<environment-configuration,this section>>) and in the web application properties file (described in <<web-application-files,this section>>) are available when processing the Jetty context XML file.

> Remember that Jetty Context XML files in the `$JETTY_BASE/webapps` directory can only refer to classes present in the environment class-path/module-path, and in the server class-path/module-path.

Reviewer 2: Also the jvm's classpath

├── environments
│   ├── {ee-current}.xml
│   └── {ee-current}-extra.xml
├── start.d
│   ├── ...
│   └── {ee-current}-ext.ini
└── webapps
    ├── wiki.properties
    └── wiki.xml
----

[[jetty-web-xml]]
== In-Application Jetty Context XML File

As discussed in <<context-xml,this section>>, the deployment of a web application can be specified with a Jetty Context XML file placed in the `$JETTY_HOME/webapps` directory.

However, the deployment of a web application can also be configured using a Jetty Context XML file placed _inside_ the web application, both for <<web-application-format-core,Core web application deployments>> and <<web-application-format-jakarta,Jakarta web application deployments>>.

This file can be useful:

> * To place all configuration inside your web application archive.

Reviewer 2: ```suggestion
* To place all configuration inside your application archive.
```

> * To perform configuration of the web application that can only occur _after_ the web application's class-path has been created, for example because it references classes present in web application itself.

Reviewer 2: ```suggestion
* To perform configuration of the web application that can only occur _after_ the application's class-path has been created, for example because it references classes present in the application itself.
```

[[jetty-web-xml-core]]

> === In-Application Jetty Context XML File for Core Web Application Deployments

Reviewer 2: ```suggestion
=== In-Application Jetty Context XML File for Core Application Deployments
```

> The file must be in the root of the web application and named `jetty-web.xml`, see <<web-application-format-core,this section>> for an example.

Reviewer 2: See previous comment - how is this being processed?

Reviewer 2: I see now that the `CoreContextHandler` is specifically looking for a file called `jetty-web.xml` and processing it. I don't like this - I know we have traditionally had a file called `jetty-web.xml` in ee web apps, but it has always been a misnomer - it appears like it's a jetty-specific extension to `web.xml` but really it's just a jetty context XML file. In addition, the expectation is that a `jetty-{ee}-web.xml` will be configuring the `WebAppContext`, not the `Server`, nor any other object (that's why the `ee` is in the name now!).  So I think the name is inappropriate in general, and very inappropriate here. The name should reflect that it is a context XML file, and that is configuring the `ContextHandler`, and ideally that it is the last configuration step before the app is started. I could live with something like `jetty-context.xml` for core apps.

Reviewer 3: The `jetty-web.xml` follows the same rules as the WebAppContext.
They configure the Context, that's it, nothing else.
The scare tactic that it can configure the Server isn't true.

https://github.com/jetty/jetty.project/blob/jetty-12.1.x/jetty-core/jetty-coreapp/src/main/java/org/eclipse/jetty/coreapp/CoreAppContext.java#L427

https://github.com/jetty/jetty.project/blob/434d1d934c9bf22c121035cce3e3302fd5fba5d6/jetty-core/jetty-coreapp/src/main/java/org/eclipse/jetty/coreapp/CoreAppContext.java#L427

This is enforced with the `XmlConfiguration.configure(this)` where the input parameter instance is the only thing that can be configured, if you specify any other class in the XML that isn't that class, it will fail.

Reviewer 3: > So I think the name is inappropriate in general, and very inappropriate here. The name should reflect that it is a context XML file, and that is configuring the `ContextHandler`, and ideally that it is the last configuration step before the app is started. I could live with something like `jetty-context.xml` for core apps.

That ship has sailed a long time ago.

Having a different name for a file in CoreAppContext that serves the same role / purpose / featureset that WebAppContext does is an even worse suggestion.

If you want a new / different name, then we need to fix it everywhere, even in WebAppContext.
But know that we cannot eliminate `jetty-web.xml` in WebAppContext, pretty much ever, as there's too many wars already built with that included now.

Reviewer 3: To me, `WEB-INF/web.xml` is a "Jakarta Web App Descriptor".
And `jetty-web.xml` is a "Jetty XML for Web Contexts".

There is no conflict.

Reviewer 2: Just because it's wrong in web apps, doesn't mean we have to perpetuate that in the brand new concept of core apps. As this is a green field, we have freedom to call the file by whatever name makes the most sense. As it has always and only been used to configure a `WebAppContext`, it makes no sense to call it `jetty-web.xml`.

`www.hostname.com`::
A fully qualified domain name. It is important to list all variants as a site may receive traffic for both `www.hostname.com` and `hostname.com`.

> `+*.hostname.com+`::

Reviewer 3: What do the plus characters do?

```
 eg: `+ blah +`
```

Author: The `+` disables inline formatting, so that e.g. the `*` is not interpreted as the start of bold, or `http://host` is not rendered as a link.
